### PR TITLE
Update to spec v0.12

### LINKED ISSUE #1182: Tracking: spec v0.12.0
## Description

This issue tracks our progress on becoming fully compliant with spec [`v0.12.0`](https://github.com/ethereum/eth2.0-specs/releases/tag/v0.12.0).

**The only issue left here is gossipsub v1.1 and it is likely to land in early July. Thankfully, 1.1 is backward-compatible with 1.0 (what we have), so this does not impede multi-client testnets.**

### Testing

- [x] The spec rewards code has been refactored to make testing easier. We'll probably have to update `ValidatorStatuses` to suit (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1747
  - https://github.com/ethereum/eth2.0-specs/pull/1826

### Beacon Chain

- [x] Adjust INACTIVITY_PENALTY_QUOTIENT to accurately reflect intention (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1712
- [x] Rename SigningRoot to SigningData (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1740
- [x] Remove PERSISTENT_ COMMITTEE_PERIOD in favor of SHARD_COMMITTEE_PERIOD ( (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1783
- [x] Disallow empty aggregation bits in attestations (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1780
- [x] Change MAX_ATTESTER_SLASHINGS from 1 to 2 (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1794
- [x] Ensure balances remain unchanged for optimal validators during inactivity leak (#1228)
  - https://github.com/ethereum/eth2.0-specs/pull/1830

### Fork choice

_Note: @paulhauner is current working to resolve these._

- [x] Add FFG-LMD vote consistency checks
  - https://github.com/ethereum/eth2.0-specs/pull/1742
- [x] [Non-substantive] Remove redundant check in validate_on_attestation 
  - https://github.com/ethereum/eth2.0-specs/pull/1755
  - IIRC we have already implemented this but we still a TODO in our code to be resolved when spec updates.
- [x] Fix our long-lasting issues with updating the justified checkpoint.

### Networking

- [x] Require "seen" aggregates to be valid to bound cache (#1236)
  - https://github.com/ethereum/eth2.0-specs/pull/1749
- [x] Clarify beacon committee assignments are not added to attnets ENR entry
  - https://github.com/ethereum/eth2.0-specs/pull/1775
- [x] Remove interop from p2p spec (#1230)
  - https://github.com/ethereum/eth2.0-specs/pull/1767
- [ ] Bump gossipsub to v1.1 and note extended validator enums for gossip conditions
  - https://github.com/ethereum/eth2.0-specs/pull/1796
  - https://github.com/ethereum/eth2.0-specs/pull/1828)
- [x] Clarify blocks in BlocksByRange must be from a single chain
  - https://github.com/ethereum/eth2.0-specs/pull/1795
- [x] Req/Resp sizes constrained by SSZ bounds (#1250)
  - https://github.com/ethereum/eth2.0-specs/pull/1800
- [x] Loosen restrictions for aggregate propagation to not require aggregator in bitfield (#1236)
  - https://github.com/ethereum/eth2.0-specs/pull/1801
- [x] Use all attestation subnets in Phase 0 (#1257)
  - https://github.com/ethereum/eth2.0-specs/pull/804
- [x] Restrict and clarify BlocksByRange request
  - https://github.com/ethereum/eth2.0-specs/pull/1835
- [x] [Verification of gossip items that aren't blocks or attestations.](https://github.com/sigp/lighthouse/blob/d79e07902e8d3daf38fb79cd43436d513ca22675/beacon_node/network/src/router/mod.rs#L290-L307) (#1253)
  - There are rules for gossip validation of exits and slashings in the spec that we haven't yet implemented, see [here](https://github.com/ethereum/eth2.0-specs/blob/v0.12.0/specs/phase0/p2p-interface.md#global-topics).

### SSZ

- [x] Fix bitfield_bytes type error
  - https://github.com/ethereum/eth2.0-specs/pull/1806
  - I don't think we need to change anything, but we should double check. Update: we didn't need to change anything, see #1228.

### BLS

- [x] Update BLS to use IETF bls draft 2 with hash-to-curve draft 7 (#1183)
  -  https://github.com/ethereum/eth2.0-specs/pull/1799

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #1182

## Proposed Changes

- [x] BLS (#1183)
- [x] State processing (#1228)
- [x] Fork choice (#1229)
- [x] Attestation verification (#1236)
- [x] Exit/slashing verification (#1253)
- [x] Networking (#1244, #1230, ..., see tracking issue #1182 for detail)
- [x] Implement v0.12.1 and bump EF tests (#1259)
- [x] Bump all the `Spec: v0.11.x` tags to `Spec: v0.12.1` (#1275)

Following the roadmap in #1182

### CODE REVIEW

**Comment by michaelsproul** [comment]
FYI: going to merge `master` back into this PR from now on rather than rebasing

File: beacon_node/eth2-libp2p/src/rpc/handler.rs [hunk 1]
```diff
@@ -393,7 +393,7 @@ where
                 let err = HandlerErr::Inbound {
                     id: inbound_id,
                     proto: *protocol,
-                    error: RPCError::ErrorResponse(*code, reason.clone()),
+                    error: RPCError::ErrorResponse(*code, reason.to_string()),
```
Comment (michaelsproul) [inline beacon_node/eth2-libp2p/src/rpc/handler.rs:396]: @AgeManning @pawanjay176 I had to redo this change during the merge, I think I got it right, but just wanted to check I didn't miss anything -- I deleted a big chunk of `inject_event` that had been deleted on `master`, and transferred the `to_string` changes that had been added there to this part of the code.

Comment (pawanjay176) [reply beacon_node/eth2-libp2p/src/rpc/handler.rs:396]: Looks correct to me.

**Comment by michaelsproul** [comment]
Waiting until `master` settles down before doing another merge. Note to self: `noise.rs` can stay deleted :+1:

**Comment by michaelsproul** [comment]
Merge it! (I can't approve my own PR haha)

**Pull request merged**